### PR TITLE
Fix labels copying value from environment variables

### PR DESCRIPTION
**Pull request by thaJeztah**
This patch fixes a bug where labels use the same behavior as `--env`, resulting
in a value to be copied from environment variables with the same name as the
label if no value is set (i.e. a simple key, no `=` sign, no value).

An earlier pull request addressed similar cases for `docker run`;
2b17f4c8a8caad552025edb05a73db683fb8a5c6, but this did not address the
same situation for (e.g.) `docker service create`.

Digging in history for this bug, I found that use of the `ValidateEnv`
function for labels was added in the original implementation of the labels feature in
https://github.com/docker/docker/commit/abb5e9a0777469e64fe2c7ecfa66ea01083d2071#diff-ae476143d40e21ac0918630f7365ed3cR34

However, the design never intended it to expand environment variables,
and use of this function was either due to either a "copy/paste" of the
equivalent `--env` flags, or a misunderstanding (the name `ValidateEnv` does
not communicate that it also expands environment variables), and the existing
`ValidateLabel` was designed for _engine_ labels (which required a value to
be set).

Following the initial implementation, other parts of the code followed
the same (incorrect) approach, therefore leading the bug to be introduced
in services as well.

This patch:

- updates the `ValidateLabel` to match the expected validation
rules (this function is no longer used since 31dc5c0a9a8bdc11c7ad335aebb753ed527caa5a),
and the daemon has its own implementation)
- corrects various locations in the code where `ValidateEnv` was used instead of `ValidateLabel`.

Before this patch:

```bash
export SOME_ENV_VAR=I_AM_SOME_ENV_VAR
docker service create --label SOME_ENV_VAR --tty --name test busybox

docker service inspect --format '{{json .Spec.Labels}}' test
{"SOME_ENV_VAR":"I_AM_SOME_ENV_VAR"}
```

After this patch:

```bash
export SOME_ENV_VAR=I_AM_SOME_ENV_VAR
docker service create --label SOME_ENV_VAR --tty --name test busybox

docker container inspect --format '{{json .Config.Labels}}' test
{"SOME_ENV_VAR":""}
```

### Add back validation for invalid label values on containers

This adds validation to `docker container run` / `docker container create`;

Validation of labels provided through flags was removed in 31dc5c0a9a8bdc11c7ad335aebb753ed527caa5a,
after the validation was changed to fix labels without values, and to prevent
labels from being expanded with environment variables in 2b17f4c8a8caad552025edb05a73db683fb8a5c6

However, now empty label names from _files_ (`--label-file`) followed different
validation rules than labels passed through `--label`.

This patch adds back minimal validation for labels passed through the command-line

Before this patch:

```bash
docker container create \
--name label \
--label==with-leading-equal-sign \
--label=without-value \
--label=somelabel=somevalue \
--label " = " \
--label=with-quotes-in-value='{"foo"}' \
--label='with"quotes"in-key=test' \
busybox

docker container inspect --format '{{json .Config.Labels}}' label
```

```json
{
"": "with-leading-equal-sign",
" ": " ",
"somelabel": "somevalue",
"with\"quotes\"in-key": "test",
"with-quotes-in-value": "{\"foo\"}",
"without-value": ""
}
```

After this patch:

```bash
docker container create \
--name label \
--label==with-leading-equal-sign \
--label=without-value \
--label=somelabel=somevalue \
--label " = " \
--label=with-quotes-in-value='{"foo"}' \
--label='with"quotes"in-key=test' \
busybox

invalid argument "=with-leading-equal-sign" for "-l, --label" flag: invalid label format: "=with-leading-equal-sign"
```

### CODE REVIEW

**Comment by thaJeztah** [comment]
@silvin-lubecki @vdemeester ptal

partially thinking if we should also move more of this validation to the daemon 🤔

**Comment by silvin-lubecki** [comment]
Were you an archeologist in a previous life @thaJeztah 😆 ? The description is impressive!

File: opts/opts.go [hunk 1]
```diff
@@ -267,10 +267,24 @@ func validateDomain(val string) (string, error) {
...
+// otherwise, so any string value is accepted for both, which includes whitespace
+// (for values) and quotes (surrounding, or embedded in key or value).
+//
+// TODO discuss if quotes (and other special characters) should be valid or invalid for keys
+// TODO discuss if leading/trailing whitespace in keys should be preserved (and valid)
 func ValidateLabel(val string) (string, error) {
-	if strings.Count(val, "=") < 1 {
-		return "", fmt.Errorf("bad attribute format: %s", val)
+	arr := strings.SplitN(val, "=", 2)
+	key := strings.TrimLeft(arr[0], whiteSpaces)
```
Comment (silvin-lubecki) [inline opts/opts.go]: Why trimming these whitespaces in the key?
It contradicts your comment:
> key must be a non-empty string, and not contain whitespaces.

File: opts/opts.go [hunk 2]
```diff
@@ -267,10 +267,24 @@ func validateDomain(val string) (string, error) {
...
+	if key == "" {
+		return "", fmt.Errorf("empty label name: '%s'", val)
```
Comment (silvin-lubecki) [inline opts/opts.go]: Isn't the label `name` the `key`? If so you should replace `val` by `key`.

Comment (silvin-lubecki) [reply opts/opts.go]: Well ok as the key is empty, there's no point printing it in the message... Then maybe rephrase it with something like: `"invalid label '%s': empty name", val` ?

File: opts/opts_test.go [hunk 3]
```diff
@@ -175,19 +177,104 @@ func TestValidateDNSSearch(t *testing.T) {
...
+			expectedErr: `empty label name: ' '`,
+		},
+		{
+			name:        "whitespace around equal-sign",
+			value:       " = ",
+			expectedErr: `empty label name: ' = '`,
+		},
+		{
+			name:  "leading whitespace",
+			value: "    label=value",
```
Comment (silvin-lubecki) [inline opts/opts_test.go]: Do we really want to validate this? Or maybe that's already something valid and trimmed on the engine side?

File: opts/opts_test.go [hunk 4]
```diff
@@ -175,19 +177,104 @@ func TestValidateDNSSearch(t *testing.T) {
...
+		},
+		{
+			name:        "whitespaces in key without value",
+			value:       "this is a label without value",
+			expectedErr: `label 'this is a label without value' has white spaces`,
+		},
+		{
+			name:        "whitespaces in key",
+			value:       "this is a label=value",
+			expectedErr: `label 'this is a label' has white spaces`,
```
Comment (silvin-lubecki) [inline opts/opts_test.go]: `has white spaces` or `has whitespaces` ? 🤔

**Comment by silvin-lubecki** [comment]
> partially thinking if we should also move more of this validation to the daemon 🤔

What do you mean? Which part do you want to move? I think the validation is simple enough on the client side 🤔

File: opts/opts.go (hunk 2 above)
Comment (thaJeztah) [reply opts/opts.go]: Yes, it's not proper english; I copied these from https://github.com/docker/cli/blob/9e712073274021ef1e552ecbb54595142b49edbb/opts/file.go#L51-L54

but perhaps I should update both

File: opts/opts.go (hunk 1 above)
Comment (thaJeztah) [reply opts/opts.go]: Hm, yeah, it may not _contain_ whitespace, but leading whitespace is removed. Basically, I copied the logic from https://github.com/docker/cli/blob/9e712073274021ef1e552ecbb54595142b49edbb/opts/file.go#L51-L54 to be consistent

**Comment by thaJeztah** [comment]
updated PTAL
